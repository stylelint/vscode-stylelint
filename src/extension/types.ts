import type TypedEventEmitter from 'typed-emitter';
import type { DidRegisterDocumentFormattingEditProviderNotificationParams } from '../server/index';

/**
 * VS Code extension event names.
 */
export type ExtensionEvents = {
	DidRegisterDocumentFormattingEditProvider: (
		params: DidRegisterDocumentFormattingEditProviderNotificationParams,
	) => void;
	DidResetConfiguration: () => void;
};

/**
 * VS Code extension public API.
 */
export type PublicApi = TypedEventEmitter<ExtensionEvents> & {
	codeActionReady: boolean;
};

/**
 * Extension API event names.
 */
export enum ApiEvent {
	DidRegisterDocumentFormattingEditProvider = 'DidRegisterDocumentFormattingEditProvider',
	DidResetConfiguration = 'DidResetConfiguration',
}
