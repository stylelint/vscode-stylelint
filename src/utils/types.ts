/**
 * Makes all properties of type `T` optional except for those in `K`.
 */
export type OptionalExcept<T extends Record<never, never>, K extends keyof T> = Pick<T, K> &
	Partial<Pick<T, K>>;

/**
 * Makes all properties of type `T` required except for those in `K`.
 */
export type RequiredExcept<T extends Record<never, never>, K extends keyof T> = Pick<
	T,
	Exclude<keyof T, K>
> &
	Required<Pick<T, K>>;

/**
 * Extracts keys from a type `T` the values of which are type `V`.
 */
export type ExtractKeysOfValueType<T, V> = {
	[K in keyof T]-?: T[K] extends V ? K : never;
}[keyof T];

/**
 * Given a function `T`, returns a function type with the same parameters, but
 * with the return type `Promise<R> | R` where `R` is the return type of `T`.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type MaybeAsync<T extends (...args: any[]) => any> = (
	...args: Parameters<T>
) => Promise<ReturnType<T>> | ReturnType<T>;

/**
 * Given a type `T`, returns a type with only the public properties of `T`.
 */
export type PublicOnly<T> = Pick<T, keyof T>;

/**
 * Given a type `T`, returns a type with only the public properties of `T` and
 * its descendants.
 */
export type PublicOnlyDeep<T> = {
	// eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
	[K in keyof T]: T[K] extends Function ? T[K] : T[K] extends object ? PublicOnlyDeep<T[K]> : T[K];
};
