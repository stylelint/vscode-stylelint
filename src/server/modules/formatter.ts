import type { TextDocument } from 'vscode-languageserver-textdocument';
import * as LSP from 'vscode-languageserver-protocol';
import { formattingOptionsToRules } from '../../utils/stylelint/index';
import { Notification } from '../types';
import type {
	LanguageServerContext,
	LanguageServerModuleConstructorParameters,
	LanguageServerModule,
} from '../types';
import type winston from 'winston';
import { URI } from 'vscode-uri';

export class FormatterModule implements LanguageServerModule {
	static id = 'formatter';

	/**
	 * The language server context.
	 */
	#context: LanguageServerContext;

	/**
	 * The logger to use, if any.
	 */
	#logger: winston.Logger | undefined;

	/**
	 * Whether or not dynamic registration for document formatting should be
	 * attempted.
	 */
	#registerDynamically = false;

	/**
	 * Promises that resolve to the disposables for the dynamically registered
	 * document formatters, by resource URI.
	 */
	#registrations = new Map<string, Promise<LSP.Disposable>>();

	/**
	 * Disposables for handlers.
	 */
	#disposables: LSP.Disposable[] = [];

	constructor({ context, logger }: LanguageServerModuleConstructorParameters) {
		this.#context = context;
		this.#logger = logger;
	}

	async #shouldFormat(document: TextDocument): Promise<boolean> {
		const options = await this.#context.getOptions(document.uri);

		return options.validate.includes(document.languageId);
	}

	dispose(): void {
		this.#context.connection.onDocumentFormatting(() => undefined);
		this.#disposables.forEach((disposable) => disposable.dispose());
		this.#disposables.length = 0;
		this.#deregisterAll();
	}

	onInitialize({ capabilities }: LSP.InitializeParams): Partial<LSP.InitializeResult> {
		this.#registerDynamically = Boolean(capabilities.textDocument?.formatting?.dynamicRegistration);

		return {
			capabilities: {
				// Use static registration if dynamic registration is not
				// supported by the client
				documentFormattingProvider: !this.#registerDynamically,
			},
		};
	}

	async #register(document: TextDocument): Promise<void> {
		if (
			!this.#registerDynamically ||
			!(await this.#shouldFormat(document)) ||
			this.#registrations.has(document.uri)
		) {
			return;
		}

		const { scheme, fsPath } = URI.parse(document.uri);

		const pattern = (scheme === 'file' ? fsPath.replace(/\\/g, '/') : fsPath).replace(
			/[[\]{}]/g,
			'?',
		);

		const filter: LSP.DocumentFilter = { scheme, pattern };
		const options: LSP.DocumentFormattingRegistrationOptions = {
			documentSelector: [filter],
		};

		this.#registrations.set(
			document.uri,
			this.#context.connection.client.register(LSP.DocumentFormattingRequest.type, options),
		);

		this.#logger?.debug('Registering formatter for document', {
			uri: document.uri,
			options,
		});

		await this.#context.connection.sendNotification(
			Notification.DidRegisterDocumentFormattingEditProvider,
			{ uri: document.uri, options },
		);
	}

	#deregister(uri: string): void {
		const registration = this.#registrations.get(uri);

		if (!registration) {
			return;
		}

		this.#logger?.debug('Deregistering formatter for document', { uri });

		registration
			.then((d) => d.dispose())
			.catch((error: unknown) => {
				this.#logger?.error('Error deregistering formatter for document', {
					uri,
					error,
				});
			});

		this.#registrations.delete(uri);
	}

	#deregisterAll(): void {
		for (const [uri, registration] of this.#registrations) {
			this.#logger?.debug('Deregistering formatter for document', { uri });

			registration
				.then((d) => d.dispose())
				.catch((error: unknown) => {
					this.#logger?.error('Error deregistering formatter for document', {
						uri,
						error,
					});
				});
		}

		this.#registrations.clear();
	}

	onDidRegisterHandlers(): void {
		this.#logger?.debug('Registering connection.onDocumentFormatting handler');
		this.#context.connection.onDocumentFormatting(async ({ textDocument, options }) => {
			this.#logger?.debug('Received onDocumentFormatting', {
				textDocument,
				options,
			});

			if (!textDocument) {
				this.#logger?.debug('No text document provided, ignoring');

				return null;
			}

			const { uri } = textDocument;
			const document = this.#context.documents.get(uri);

			if (!document || !(await this.#shouldFormat(document))) {
				if (this.#logger?.isDebugEnabled()) {
					if (!document) {
						this.#logger.debug('Unknown document, ignoring', { uri });
					} else {
						this.#logger.debug('Document should not be formatted, ignoring', {
							uri,
							language: document.languageId,
						});
					}
				}

				return null;
			}

			const linterOptions = {
				config: {
					rules: formattingOptionsToRules(options),
				},
			};

			this.#logger?.debug('Formatting document', { uri, linterOptions });

			const fixes = this.#context.getFixes(document, linterOptions);

			this.#logger?.debug('Returning fixes', { uri, fixes });

			return fixes;
		});
		this.#logger?.debug('connection.onDocumentFormatting handler registered');

		this.#logger?.debug('Registering documents.onDidOpen handler');
		this.#disposables.push(
			this.#context.documents.onDidOpen(({ document }) => this.#register(document)),
		);
		this.#logger?.debug('documents.onDidOpen handler registered');

		this.#logger?.debug('Registering documents.onDidChangeContent handler');
		this.#disposables.push(
			this.#context.documents.onDidChangeContent(({ document }) => this.#register(document)),
		);
		this.#logger?.debug('documents.onDidChangeContent handler registered');

		this.#logger?.debug('Registering documents.onDidSave handler');
		this.#disposables.push(
			this.#context.documents.onDidSave(({ document }) => this.#register(document)),
		);
		this.#logger?.debug('documents.onDidSave handler registered');

		this.#logger?.debug('Registering documents.onDidClose handler');
		this.#disposables.push(
			this.#context.documents.onDidClose(({ document }) => this.#deregister(document.uri)),
		);
		this.#logger?.debug('documents.onDidClose handler registered');

		this.#logger?.debug('Registering DidChangeConfigurationNotification');
		this.#disposables.push(
			this.#context.notifications.on(LSP.DidChangeConfigurationNotification.type, () =>
				this.#deregisterAll(),
			),
		);
		this.#logger?.debug('DidChangeConfigurationNotification registered');

		this.#logger?.debug('Registering DidChangeWorkspaceFoldersNotification');
		this.#disposables.push(
			this.#context.notifications.on(LSP.DidChangeWorkspaceFoldersNotification.type, () =>
				this.#deregisterAll(),
			),
		);
		this.#logger?.debug('DidChangeWorkspaceFoldersNotification registered');
	}
}
