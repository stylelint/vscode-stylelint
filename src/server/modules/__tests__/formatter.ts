jest.mock('vscode-uri', () => {
	const vscodeUri = jest.requireActual('vscode-uri');
	const mock = jest.createMockFromModule<typeof vscodeUri>('vscode-uri');

	mock.URI.parse = (uri: string) => {
		const parsed = vscodeUri.URI.parse(uri);

		if (parsed.scheme === 'file') {
			return parsed;
		}

		const fsPath = parsed.fsPath?.replace(/\//g, '\\');

		return {
			...parsed,
			fsPath,
		};
	};

	return mock;
});

import { WorkDoneProgressReporter } from 'vscode-languageserver';
import * as LSP from 'vscode-languageserver-protocol';
import { TextDocument } from 'vscode-languageserver-textdocument';
import { Position, TextEdit } from 'vscode-languageserver-types';

import { Notification } from '../../types';
import { FormatterModule } from '../formatter';

let mockContext = serverMocks.getContext();
let mockLogger = serverMocks.getLogger();

describe('FormatterModule', () => {
	beforeEach(() => {
		jest.restoreAllMocks();
		mockContext = serverMocks.getContext();
		mockContext.__options.validate = [];
		mockLogger = serverMocks.getLogger();
	});

	test('should be constructable', () => {
		expect(() => new FormatterModule({ context: mockContext.__typed() })).not.toThrow();
	});

	test('without client dynamic registration support, onInitialize should request static registration', () => {
		const module = new FormatterModule({ context: mockContext.__typed() });

		expect(
			module.onInitialize({
				capabilities: {
					textDocument: {
						formatting: { dynamicRegistration: false },
					},
				},
			} as unknown as LSP.InitializeParams),
		).toMatchSnapshot();
	});

	test('with client dynamic registration support, onInitialize should not request static registration', () => {
		const module = new FormatterModule({ context: mockContext.__typed() });

		expect(
			module.onInitialize({
				capabilities: {
					textDocument: {
						formatting: { dynamicRegistration: true },
					},
				},
			} as unknown as LSP.InitializeParams),
		).toMatchSnapshot();
	});

	test('onDidRegisterHandlers should register a document formatting command handler', () => {
		const module = new FormatterModule({ context: mockContext.__typed() });

		module.onDidRegisterHandlers();

		expect(mockContext.connection.onDocumentFormatting).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.onDocumentFormatting).toHaveBeenCalledWith(expect.any(Function));
	});

	test('should format documents', async () => {
		mockContext.documents.get.mockReturnValue(TextDocument.create('foo', 'bar', 1, 'a {}'));
		mockContext.__options.validate = ['bar'];
		mockContext.getFixes.mockResolvedValue([TextEdit.insert(Position.create(0, 0), 'text')]);

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onDocumentFormatting.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'foo' },
				options: {
					insertSpaces: true,
					tabSize: 2,
					insertFinalNewline: true,
					trimFinalNewlines: false,
				},
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toMatchSnapshot();
		expect(mockContext.getFixes.mock.calls[0]).toMatchSnapshot();
		expect(await mockContext.getFixes.mock.results[0].value).toStrictEqual(result);
	});

	test('with no text document, should not attempt to format', async () => {
		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onDocumentFormatting.mock.calls[0][0];

		const result = await handler(
			{
				options: { insertSpaces: true, tabSize: 2 },
				textDocument: undefined as unknown as LSP.TextDocumentIdentifier,
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toBeNull();
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('No text document provided, ignoring');
	});

	test('if no matching document exists, should not attempt to format', async () => {
		mockContext.documents.get.mockReturnValue(undefined);

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onDocumentFormatting.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'foo' },
				options: { insertSpaces: true, tabSize: 2 },
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toBeNull();
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Unknown document, ignoring', { uri: 'foo' });
	});

	test('if document language ID is not in options, should not attempt to format', async () => {
		mockContext.documents.get.mockReturnValue(TextDocument.create('foo', 'bar', 1, 'a {}'));
		mockContext.__options.validate = ['baz'];

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onDocumentFormatting.mock.calls[0][0];

		const result = await handler(
			{
				textDocument: { uri: 'foo' },
				options: { insertSpaces: true, tabSize: 2 },
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toBeNull();
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Document should not be formatted, ignoring',
			{
				uri: 'foo',
				language: 'bar',
			},
		);
	});

	test('with no debug log level and no valid document, should not attempt to log reason', async () => {
		mockContext.documents.get.mockReturnValue(TextDocument.create('foo', 'bar', 1, 'a {}'));
		mockContext.__options.validate = ['baz'];
		mockLogger.isDebugEnabled.mockReturnValue(false);

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onDocumentFormatting.mock.calls[0][0];

		const handlerParams: LSP.DocumentFormattingParams = {
			textDocument: { uri: 'foo' },
			options: { insertSpaces: true, tabSize: 2 },
		};

		const result = await handler(
			handlerParams,
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toBeNull();
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Received onDocumentFormatting',
			handlerParams,
		);
	});

	test("without client dynamic registration support, documents.onDidOpen shouldn't register a formatter", async () => {
		mockContext.__options.validate = ['bar'];

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: false },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await onDidOpenHandler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.connection.client.register).not.toHaveBeenCalled();
	});

	test("without client dynamic registration support, documents.onDidChangeContent shouldn't register a formatter", async () => {
		mockContext.__options.validate = ['bar'];

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: false },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidChangeContentHandler = mockContext.documents.onDidChangeContent.mock.calls[0][0];

		await onDidChangeContentHandler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.connection.client.register).not.toHaveBeenCalled();
	});

	test("without client dynamic registration support, documents.onDidSave shouldn't register a formatter", async () => {
		mockContext.__options.validate = ['bar'];

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: false },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidSaveHandler = mockContext.documents.onDidSave.mock.calls[0][0];

		await onDidSaveHandler({
			document: TextDocument.create('foo', 'bar', 1, 'a {}'),
		});

		expect(mockContext.connection.client.register).not.toHaveBeenCalled();
	});

	test('with client dynamic registration support, documents.onDidOpen should register a formatter', async () => {
		mockContext.__options.validate = ['bar'];

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await onDidOpenHandler({
			document: TextDocument.create('file:///dir/test.css', 'bar', 1, 'a {}'),
		});
		await onDidOpenHandler({
			document: TextDocument.create('scheme:///dir/test.css', 'bar', 1, 'a {}'),
		});

		expect(mockContext.connection.client.register).toHaveBeenCalledWith(
			LSP.DocumentFormattingRequest.type,
			{ documentSelector: [{ scheme: 'file', pattern: '/dir/test.css' }] },
		);
		expect(mockContext.connection.client.register).toHaveBeenCalledWith(
			LSP.DocumentFormattingRequest.type,
			{ documentSelector: [{ scheme: 'scheme', pattern: '\\dir\\test.css' }] },
		);
	});

	test('with client dynamic registration support, documents.onDidChangeContent should register a formatter', async () => {
		mockContext.__options.validate = ['bar'];

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidChangeContentHandler = mockContext.documents.onDidChangeContent.mock.calls[0][0];

		await onDidChangeContentHandler({
			document: TextDocument.create('file:///dir/test.css', 'bar', 1, 'a {}'),
		});

		expect(mockContext.connection.client.register).toHaveBeenCalledWith(
			LSP.DocumentFormattingRequest.type,
			{ documentSelector: [{ scheme: 'file', pattern: '/dir/test.css' }] },
		);
	});

	test('with client dynamic registration support, documents.onDidSave should register a formatter', async () => {
		mockContext.__options.validate = ['bar'];

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidSaveHandler = mockContext.documents.onDidSave.mock.calls[0][0];

		await onDidSaveHandler({
			document: TextDocument.create('file:///dir/test.css', 'bar', 1, 'a {}'),
		});

		expect(mockContext.connection.client.register).toHaveBeenCalledWith(
			LSP.DocumentFormattingRequest.type,
			{ documentSelector: [{ scheme: 'file', pattern: '/dir/test.css' }] },
		);
	});

	test('when a formatter was registered, documents.onDidClose should dispose the old registration', async () => {
		mockContext.__options.validate = ['bar'];
		mockLogger.isDebugEnabled.mockReturnValue(true);

		const mockRegistration = { dispose: jest.fn() };

		mockContext.connection.client.register.mockResolvedValueOnce(mockRegistration);

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const document = TextDocument.create('file:///dir/test.css', 'bar', 1, 'a {}');

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];
		const onDidCloseHandler = mockContext.documents.onDidClose.mock.calls[0][0];

		await onDidOpenHandler({ document });
		await onDidCloseHandler({ document });

		expect(mockRegistration.dispose).toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenCalledWith('Deregistering formatter for document', {
			uri: 'file:///dir/test.css',
		});
	});

	test('when a formatter was registered, if registration rejects, documents.onDidClose should log an error', async () => {
		const error = new Error('test');

		mockContext.__options.validate = ['bar'];
		mockLogger.isDebugEnabled.mockReturnValue(true);

		mockContext.connection.client.register.mockRejectedValueOnce(error);

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const document = TextDocument.create('file:///dir/test.css', 'bar', 1, 'a {}');

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];
		const onDidCloseHandler = mockContext.documents.onDidClose.mock.calls[0][0];

		await onDidOpenHandler({ document });
		await onDidCloseHandler({ document });

		await new Promise((resolve) => setImmediate(resolve));

		expect(mockLogger.error).toHaveBeenCalledWith('Error deregistering formatter for document', {
			uri: 'file:///dir/test.css',
			error,
		});
	});

	test('when a formatter was not registered, documents.onDidClose should not try to dispose a registration', async () => {
		mockContext.__options.validate = ['bar'];
		mockLogger.isDebugEnabled.mockReturnValue(true);

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidCloseHandler = mockContext.documents.onDidClose.mock.calls[0][0];

		await onDidCloseHandler({
			document: TextDocument.create('file:///dir/test.css', 'bar', 1, 'a {}'),
		});

		expect(mockLogger.debug).not.toHaveBeenCalledWith('Deregistering formatter for document', {
			uri: 'file:///dir/test.css',
		});
	});

	test('when formatters were registered, DidChangeConfigurationNotification should deregister all registrations', async () => {
		mockContext.__options.validate = ['bar'];
		mockLogger.isDebugEnabled.mockReturnValue(true);

		const mockRegistration = { dispose: jest.fn() };

		mockContext.connection.client.register.mockResolvedValue(mockRegistration);

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await onDidOpenHandler({
			document: TextDocument.create('file:///dir/test-1.css', 'bar', 1, 'a {}'),
		});
		await onDidOpenHandler({
			document: TextDocument.create('file:///dir/test-2.css', 'bar', 1, 'a {}'),
		});

		const didChangeConfigurationNotificationHandler = mockContext.notifications.on.mock.calls.find(
			([type]) => type === LSP.DidChangeConfigurationNotification.type,
		)?.[1];

		await didChangeConfigurationNotificationHandler?.();

		expect(mockRegistration.dispose).toHaveBeenCalledTimes(2);
	});

	test('when formatters were registered, if registration rejects, DidChangeConfigurationNotification should log an error', async () => {
		const error = new Error('test');

		mockContext.__options.validate = ['bar'];
		mockLogger.isDebugEnabled.mockReturnValue(true);

		mockContext.connection.client.register.mockRejectedValue(error);

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await onDidOpenHandler({
			document: TextDocument.create('file:///dir/test-1.css', 'bar', 1, 'a {}'),
		});
		await onDidOpenHandler({
			document: TextDocument.create('file:///dir/test-2.css', 'bar', 1, 'a {}'),
		});

		const didChangeConfigurationNotificationHandler = mockContext.notifications.on.mock.calls.find(
			([type]) => type === LSP.DidChangeConfigurationNotification.type,
		)?.[1];

		await didChangeConfigurationNotificationHandler?.();

		await new Promise((resolve) => setImmediate(resolve));

		expect(mockLogger.error).toHaveBeenCalledWith('Error deregistering formatter for document', {
			uri: 'file:///dir/test-1.css',
			error,
		});
		expect(mockLogger.error).toHaveBeenCalledWith('Error deregistering formatter for document', {
			uri: 'file:///dir/test-2.css',
			error,
		});
	});

	test('when formatters were registered, DidChangeWorkspaceFoldersNotification should deregister all registrations', async () => {
		mockContext.__options.validate = ['bar'];
		mockLogger.isDebugEnabled.mockReturnValue(true);

		const mockRegistration = { dispose: jest.fn() };

		mockContext.connection.client.register.mockResolvedValue(mockRegistration);

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await onDidOpenHandler({
			document: TextDocument.create('file:///dir/test-1.css', 'bar', 1, 'a {}'),
		});
		await onDidOpenHandler({
			document: TextDocument.create('file:///dir/test-2.css', 'bar', 1, 'a {}'),
		});

		const didChangeWorkspaceFoldersNotificationHandler =
			mockContext.notifications.on.mock.calls.find(
				([type]) => type === LSP.DidChangeWorkspaceFoldersNotification.type,
			)?.[1];

		await didChangeWorkspaceFoldersNotificationHandler?.();

		expect(mockRegistration.dispose).toHaveBeenCalledTimes(2);
	});

	test('when formatters were registered, if registration rejects, DidChangeWorkspaceFoldersNotification should log an error', async () => {
		const error = new Error('test');

		mockContext.__options.validate = ['bar'];
		mockLogger.isDebugEnabled.mockReturnValue(true);

		mockContext.connection.client.register.mockRejectedValue(error);

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await onDidOpenHandler({
			document: TextDocument.create('file:///dir/test-1.css', 'bar', 1, 'a {}'),
		});
		await onDidOpenHandler({
			document: TextDocument.create('file:///dir/test-2.css', 'bar', 1, 'a {}'),
		});

		const didChangeWorkspaceFoldersNotificationHandler =
			mockContext.notifications.on.mock.calls.find(
				([type]) => type === LSP.DidChangeWorkspaceFoldersNotification.type,
			)?.[1];

		await didChangeWorkspaceFoldersNotificationHandler?.();

		await new Promise((resolve) => setImmediate(resolve));

		expect(mockLogger.error).toHaveBeenCalledWith('Error deregistering formatter for document', {
			uri: 'file:///dir/test-1.css',
			error,
		});
		expect(mockLogger.error).toHaveBeenCalledWith('Error deregistering formatter for document', {
			uri: 'file:///dir/test-2.css',
			error,
		});
	});

	test('with client dynamic registration support, only one formatter should be registered per document', async () => {
		mockContext.__options.validate = ['bar'];

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const document = TextDocument.create('file:///dir/test.css', 'bar', 1, 'a {}');

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];
		const onDidSaveHandler = mockContext.documents.onDidSave.mock.calls[0][0];

		await onDidOpenHandler({ document });
		await onDidSaveHandler({ document });

		expect(mockContext.connection.client.register).toHaveBeenCalledTimes(1);
	});

	test('when a formatter is registered, a notification should be sent', async () => {
		mockContext.__options.validate = ['bar'];

		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await onDidOpenHandler({
			document: TextDocument.create('file:///dir/test.css', 'bar', 1, 'a {}'),
		});

		expect(mockContext.connection.sendNotification).toHaveBeenCalledWith(
			Notification.DidRegisterDocumentFormattingEditProvider,
			{
				uri: 'file:///dir/test.css',
				options: {
					documentSelector: [{ scheme: 'file', pattern: '/dir/test.css' }],
				},
			},
		);
	});

	it('should be disposable', () => {
		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		expect(module).toHaveProperty('dispose');
		expect(module.dispose).toBeInstanceOf(Function);
	});

	it('should set a no-op formatting handler when disposed', async () => {
		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();
		module.dispose();

		const handler = mockContext.connection.onDocumentFormatting.mock.calls[1][0];

		const result = await handler(
			{
				textDocument: { uri: 'file:///dir/test.css' },
				options: {
					tabSize: 4,
					insertSpaces: true,
				},
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(mockContext.connection.onDocumentFormatting).toHaveBeenCalledTimes(2);
		expect(result).toBeUndefined();
	});

	it('should dispose all handler registrations when disposed', () => {
		const module = new FormatterModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();
		module.dispose();

		const disposables = [
			...mockContext.notifications.on.mock.results,
			...mockContext.commands.on.mock.results,
			...mockContext.documents.onDidOpen.mock.results,
			...mockContext.documents.onDidSave.mock.results,
			...mockContext.documents.onDidClose.mock.results,
			...mockContext.documents.onDidChangeContent.mock.results,
		];

		expect(disposables).toHaveLength(6);

		for (const disposable of disposables) {
			expect(disposable.value.dispose).toHaveBeenCalledTimes(1);
		}
	});
});
