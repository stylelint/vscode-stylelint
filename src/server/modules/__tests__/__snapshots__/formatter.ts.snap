// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormatterModule should format documents 1`] = `
[
  {
    "newText": "text",
    "range": {
      "end": {
        "character": 0,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
  },
]
`;

exports[`FormatterModule should format documents 2`] = `
[
  FullTextDocument {
    "_content": "a {}",
    "_languageId": "bar",
    "_lineOffsets": undefined,
    "_uri": "foo",
    "_version": 1,
  },
  {
    "config": {
      "rules": {
        "indentation": [
          2,
        ],
        "no-missing-end-of-source-newline": true,
      },
    },
  },
]
`;

exports[`FormatterModule with client dynamic registration support, onInitialize should not request static registration 1`] = `
{
  "capabilities": {
    "documentFormattingProvider": false,
  },
}
`;

exports[`FormatterModule without client dynamic registration support, onInitialize should request static registration 1`] = `
{
  "capabilities": {
    "documentFormattingProvider": true,
  },
}
`;
